File: kiosk/src/lib/rfid.ts
/**
 * The house RFID format, on the kiosk's side of the wire.
 *
 * A tag is stored as 24 characters, zero-padded on the left
 * (`000000000000000000100348`), so a handheld reading "100348" and a
 * fixed reader reading the tag's full EPC land on one value —
 * `displayRfid` (`@portal/lib/format`) strips that padding again for
 * display only. RFID Enroll pads here so the operator can see the exact
 * value that will be stored before they commit to it; the endpoint
 * normalizes identically and independently (`normalize_rfid` in
 * api/.../routes/kiosk.py), because nothing the kiosk sends is trusted.
 *
 * Whitespace goes everywhere, not just at the ends: a reader may
 * space-separate an EPC's words, and those spaces are formatting, not
 * part of the tag.
 */

export const RFID_LENGTH = 24;

/** Why a value is not a tag: nothing was scanned, the reader produced
 *  something that is not a plain alphanumeric tag, or it is longer than
 *  the stored format can hold. */
export type RfidProblem = 'empty' | 'not_alphanumeric' | 'too_long';

export interface PaddedRfid {
  /** The value as it would be stored, or null when `problem` is set. */
  tag: string | null;
  problem: RfidProblem | null;
}

const ALPHANUMERIC = /^[0-9A-Z]+$/;

export function padRfid(raw: string): PaddedRfid {
  const tag = raw.replace(/\s+/g, '').toUpperCase();
  if (!tag) return { tag: null, problem: 'empty' };
  if (!ALPHANUMERIC.test(tag)) return { tag: null, problem: 'not_alphanumeric' };
  if (tag.length > RFID_LENGTH) return { tag: null, problem: 'too_long' };
  return { tag: tag.padStart(RFID_LENGTH, '0'), problem: null };
}

/** What a rejected value says out loud on the Enroll screen. */
export function rfidProblemText(problem: RfidProblem): string {
  switch (problem) {
    case 'empty': return 'Scan the RFID tag.';
    case 'too_long': return `That tag is longer than ${RFID_LENGTH} characters.`;
    default: return "That tag has characters we can't store — letters and numbers only.";
  }
}

File: Android_Kiosk_App/app/src/main/java/com/serversherpa/kiosk/core/scan/Rfid.kt
package com.serversherpa.kiosk.core.scan

/** The house RFID format: 24 characters, zero-padded on the left. The
 *  server normalizes independently (`normalize_rfid` in routes/kiosk.py);
 *  the kiosk pads so the operator sees what will be stored. */
const val RFID_LENGTH = 24

enum class RfidProblem { EMPTY, NOT_ALPHANUMERIC, TOO_LONG }

data class PaddedRfid(val tag: String?, val problem: RfidProblem?)

private val ALPHANUMERIC = Regex("^[0-9A-Z]+$")

fun padRfid(raw: String): PaddedRfid {
    val tag = raw.replace(Regex("\\s+"), "").uppercase()
    if (tag.isEmpty()) return PaddedRfid(null, RfidProblem.EMPTY)
    if (!ALPHANUMERIC.matches(tag)) return PaddedRfid(null, RfidProblem.NOT_ALPHANUMERIC)
    if (tag.length > RFID_LENGTH) return PaddedRfid(null, RfidProblem.TOO_LONG)
    return PaddedRfid(tag.padStart(RFID_LENGTH, '0'), null)
}

fun rfidProblemText(problem: RfidProblem): String = when (problem) {
    RfidProblem.EMPTY -> "Scan the RFID tag."
    RfidProblem.TOO_LONG -> "That tag is longer than $RFID_LENGTH characters."
    RfidProblem.NOT_ALPHANUMERIC -> "That tag has characters we can't store — letters and numbers only."
}

/** portal/src/lib/format.ts: leading zeros stripped for display; a
 *  missing tag renders as an em dash. */
fun displayRfid(tag: String?): String {
    if (tag.isNullOrEmpty()) return "—"
    return tag.replace(Regex("^0+(?=.)"), "")
}
